cmd/prometheus/main.go

```diff
@@ -282,104 +278,182 @@ func main() {
 		notifier,
 	}
 
-	// Wait for reload or termination signals. Start the handler for SIGHUP as
-	// early as possible, but ignore it until we are ready to handle reloading
-	// our config.
-	go func() {
-		<-hupReady
-		for {
-			select {
-			case <-hup:
-				if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-					level.Error(logger).Log("msg", "Error reloading config", "err", err)
-				}
-			case rc := <-webHandler.Reload():
-				if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-					level.Error(logger).Log("msg", "Error reloading config", "err", err)
-					rc <- err
-				} else {
-					rc <- nil
-				}
-			}
-		}
-	}()
-
-	// Start all components while we wait for TSDB to open but only load
-	// initial config and mark ourselves as ready after it completed.
-	dbOpen := make(chan struct{})
-
-	go func() {
-		defer close(dbOpen)
-
-		level.Info(logger).Log("msg", "Starting TSDB")
+	prometheus.MustRegister(configSuccess)
+	prometheus.MustRegister(configSuccessTime)
 
-		db, err := tsdb.Open(
-			cfg.localStoragePath,
-			log.With(logger, "component", "tsdb"),
-			prometheus.DefaultRegisterer,
-			&cfg.tsdb,
+	var g group.Group
+	{
```

Comment:
I believe the idea was to use one lexical scope per actor, not one for all actors. This would enable you to use the same variable name for interrupt channels instead of the current s1, s2, ... pattern. Channels like `hupReady` which are used in multiple actors would need to be defined in the scope of the main function.

Reply:
the original implementation was with a single channel, but if you read peterbourgon's original comments few commits behind you will see why we need a separate channel as a interruptor.

Reply:
You'd still have separate channels. He talks about the advantage of using individual lexical groups in his talk.

```go
{
    cancel := make(chan struct{})
    g.Add(
        // ...
    )
}
{
    cancel := make(chan struct{})
    g.Add(
        // ...
    )
}
// ... other groups
```

Reply:
changed and it looks better like this.